x-pack/plugins/security/public/account_management/user_profile/use_update_user_profile.tsx

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { EuiButton, EuiFlexGroup, EuiFlexItem } from '@elastic/eui';
+import React, { useCallback, useRef, useState } from 'react';
+import useObservable from 'react-use/lib/useObservable';
+
+import type { NotificationsStart, ToastInput, ToastOptions } from '@kbn/core/public';
+import { i18n } from '@kbn/i18n';
+import { toMountPoint } from '@kbn/kibana-react-plugin/public';
+
+import type { UserProfileData } from './user_profile';
+import type { UserProfileAPIClient } from './user_profile_api_client';
+
+interface Deps {
+  apiClient: UserProfileAPIClient;
+  notifications: NotificationsStart;
+}
+
+interface Props {
+  notificationSuccess?: {
+    /** Flag to indicate if a notification is shown after update. Default: `true` */
+    enabled?: boolean;
+    /** Customize the title of the notification */
+    title?: string;
+    /** Customize the "page reload needed" text of the notification */
+    pageReloadText?: string;
+  };
+  /** Predicate to indicate if the update requires a page reload */
+  pageReloadChecker?: (
+    previsous: UserProfileData | null | undefined,
+    next: UserProfileData
+  ) => boolean;
+}
+
+export type UpdateUserProfileHook = (props?: Props) => {
+  /** Update the user profile */
+  update: (data: UserProfileData) => void;
+  /** Handler to show a notification after the user profile has been updated */
+  showSuccessNotification: (props: { isRefreshRequired: boolean }) => void;
+  /** Flag to indicate if currently updating */
+  isLoading: boolean;
+  /** The current user profile data */
+  userProfileData?: UserProfileData | null;
+};
+
+let useUpdateUserProfile: UpdateUserProfileHook | undefined;
+
+const i18nTexts = {
+  notificationSuccess: {
+    title: i18n.translate('xpack.security.accountManagement.userProfile.submitSuccessTitle', {
+      defaultMessage: 'Profile updated',
+    }),
+    pageReloadText: i18n.translate(
+      'xpack.security.accountManagement.userProfile.requiresPageReloadToastDescription',
+      {
+        defaultMessage: 'One or more settings require you to reload the page to take effect.',
+      }
+    ),
+  },
+};
+
+export const getUseUpdateUserProfile = (
+  { apiClient, notifications }: Deps,
+  useCacheInstance = true
+) => {
+  if (useUpdateUserProfile && useCacheInstance) {
+    return useUpdateUserProfile;
+  }
```

Comment:
What is the purpose of caching this hook and turning it into a singleton?

Reply:
So we can call the function from inside the component:
https://github.com/elastic/kibana/pull/161214/files#diff-f8e7bc19afaf6d981dd4971618f887d5b4600fa1e9dd0035859363ec3ca1e049R730

Reply:
You can call a function/hook inside a component without that function being a singleton. 

I'm trying to understand what the purpose of that behaviour is.

Reply:
Yeah I just test it an indeed React does the memoization of the hook. I didn't know that!
I removed the cache behaviour in https://github.com/elastic/kibana/pull/161214/commits/ea7acb5cf09273a8c474726b5f5dead199edbb74

Thanks for pointing it out 👍